rootston/desktop.c

```diff
@@ -402,20 +411,18 @@ struct roots_subsurface *subsurface_create(struct roots_view *view,
 	return subsurface;
 }
 
-void view_finish(struct roots_view *view) {
-	view_damage_whole(view);
-	wl_signal_emit(&view->events.destroy, view);
-
-	wl_list_remove(&view->new_subsurface.link);
-
-	struct roots_view_child *child, *tmp;
-	wl_list_for_each_safe(child, tmp, &view->children, link) {
-		child->destroy(child);
+void view_destroy(struct roots_view *view) {
+	if (view == NULL) {
+		return;
 	}
 
-	if (view->fullscreen_output) {
-		view->fullscreen_output->fullscreen_view = NULL;
+	if (view->wlr_surface != NULL) {
+		view_unmap(view);
 	}
+
+	wl_signal_emit(&view->events.destroy, view);
```

Comment:
destroy event normally goes first before all the extra cleanup is done.

Reply:
I don't want to call `view_unmap` after emitting the destroy event because the destroy event may have free'd stuff used by `view_unmap`.

Reply:
That shouldn't be an issue (and if it is, we have bigger problems).

According to our patterns, the view should have ownership over all its own properties and should be responsible for freeing them. If another part of the program frees the properties, then we'll get double frees when the view is destroyed by the `view_destroy` method. This is an application bug.

In the destroy event, the wrappers of the view should *only* be responsible for cleanup up their own wrappers and not touch the views own properties at all. Users of the view as a property should simply set it to NULL. That should be the extent of the cleanup done in those destroy handlers (and anything more is an application bug).